keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+in the form of:
+- <package>: <date> - <current test coverage>
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `pkg/scheduler`: `25 Nov 2022` - `91.0%`
+- `pkg/cache`: `25 Nov 2022` - `83.1%`
+- `pkg/controller/workload/job`: `25 Nov 2022` - `0%`
+
+#### Integration tests
+
+The following scenarios will be covered with integration tests:
+- Workload created for a Job with `kueue.x-k8s.io/startup-requirements` has `.spec.startupRequirements=QuotaAndPodsReady`
+- No workload is admitted if the `.spec.startupRequirements=QuotaAndPodsReady` is not in the `PodsReady`  condition
+- workloads are admitted if the `.spec.startupRequirements=QuotaAndPodsReady` is in the `PodsReady`  condition
+- a workload with `.spec.startupRequirements=QuotaAndPodsReady` does not get admitted if there is another admitted workload which is not in the `PodsReady`  condition
+- a workload with `.spec.startupRequirements=QuotaAndPodsReady` gets admitted if all other admitted workloads are in the `PodsReady`  condition
+- a workload which exceeds the `kueue.x-k8s.io/wait-for-pods-ready-seconds` timeout is suspended and put into the `inadmissibleWorkloads` list
+
+<!--
+Describe what tests will be added to ensure proper quality of the enhancement.
+
+After the implementation PR is merged, add the names of the tests here.
+-->
+
+### Graduation Criteria
+
+<!--
+
+Clearly define what it means for the feature to be implemented and
+considered stable.
+
+If the feature you are introducing has high complexity, consider adding graduation
+milestones with these graduation criteria:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+-->
+
+N/A
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+Delaying of workload admission until all pods are scheduled may decrease
+thoughput significantly. Especially, if there is enough resource capacity to
+which could be otherwise used to start multiple jobs at the same time.
+
+## Alternatives
```

Comment:
Have you considered central configuration? I mean some kind of ensure-pods-start=true, but not on the Job, but on ClusterQueue or some central Kueue config (is there such a place?). Central is possibly simpler, easier for Batch admins and easier for CRD <> Workload integrations. Do we need the granularity of every job deciding differently?

Reply:
Yes, considered other options. Summarized the pros & cons in the remarks section now. I started off with Job, as it really depends on the job specificity if it requires all pods to be running at the same time. Setting it on queues would still require similar level of complexity from the implementation POV, so didn't see a point. However, there is a complexity added by introducing new Job annotation. As the contract on the objects to create workloads from gets bigger. Thus, the implementation in the Kueue's Job Reconciler would need to be also repeated for other CRDs, which could be avoided by the use of LocalQueues.